sui-execution/latest/sui-move-natives/src/config.rs

```diff
@@ -0,0 +1,185 @@
+        .ok_or_else(|| {
+            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                .with_message("gas cost is not set".to_string())
+        })?;
+    // Charge base fee
+    native_charge_gas_early_exit!(context, config_read_setting_impl_cost_base);
+
+    let value_ty = ty_args.pop().unwrap();
+    let setting_data_value_ty = ty_args.pop().unwrap();
+    let setting_value_ty = ty_args.pop().unwrap();
+    let field_setting_ty = ty_args.pop().unwrap();
+
+    let current_epoch = pop_arg!(args, u64);
+    let name_df_addr = pop_arg!(args, AccountAddress);
+    let config_addr = pop_arg!(args, AccountAddress);
+
+    let field_setting_tag: StructTag = match context.type_to_type_tag(&field_setting_ty)? {
+        TypeTag::Struct(s) => *s,
+        _ => {
+            return Err(
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Sui verifier guarantees this is a struct".to_string()),
+            )
+        }
+    };
+    let Some(field_setting_layout_opt) = context.type_to_type_layout(&field_setting_ty)? else {
+        return Ok(NativeResult::err(
+            context.gas_used(),
+            E_BCS_SERIALIZATION_FAILURE,
+        ));
+    };
+    let object_runtime: &mut ObjectRuntime = context.extensions_mut().get_mut();
+
+    let read_value_opt = consistent_value_before_current_epoch(
+        object_runtime,
+        &field_setting_ty,
+        field_setting_tag,
+        &field_setting_layout_opt,
+        &setting_value_ty,
+        &setting_data_value_ty,
+        &value_ty,
+        config_addr,
+        name_df_addr,
+        current_epoch,
+    )?;
+
+    native_charge_gas_early_exit!(
+        context,
+        config_read_setting_impl_cost_per_byte * u64::from(read_value_opt.legacy_size()).into()
+    );
+
+    Ok(NativeResult::ok(
+        context.gas_used(),
+        smallvec![read_value_opt],
+    ))
+}
+
+fn consistent_value_before_current_epoch(
+    object_runtime: &mut ObjectRuntime,
+    field_setting_ty: &Type,
+    field_setting_tag: StructTag,
+    field_setting_layout: &R::MoveTypeLayout,
+    _setting_value_ty: &Type,
+    setting_data_value_ty: &Type,
+    value_ty: &Type,
+    config_addr: AccountAddress,
+    name_df_addr: AccountAddress,
+    current_epoch: u64,
+) -> PartialVMResult<Value> {
+    let Some(field) = object_runtime.config_setting_unsequenced_read(
+        config_addr.into(),
+        name_df_addr.into(),
+        field_setting_ty,
+        field_setting_layout,
+        &MoveObjectType::from(field_setting_tag),
+    ) else {
+        return option_none(&value_ty);
+    };
+
+    let [_id, _name, setting]: [Value; 3] = unpack_struct(field)?;
+    let [data_opt]: [Value; 1] = unpack_struct(setting)?;
+    let data = match unpack_option(data_opt, &setting_data_value_ty)? {
+        None => {
+            // invariant violation?
```

Comment:
Need to think on it a bit more, but I think an invariant violation here makes sense.

Reply:
I am just very very scared of doing that here. Maybe just log it?

Reply:
That's fine with me as well :) IMO this should either be a log or a panic, and I'm fine with either